azure/services/aksextensions/spec.go

```diff
@@ -0,0 +1,110 @@
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package aksextensions
+
+import (
+	"context"
+
+	asokubernetesconfigurationv1 "github.com/Azure/azure-service-operator/v2/api/kubernetesconfiguration/v1api20230501"
+	"github.com/Azure/azure-service-operator/v2/pkg/genruntime"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/utils/ptr"
+	infrav1 "sigs.k8s.io/cluster-api-provider-azure/api/v1beta1"
+)
+
+// AKSExtensionSpec defines the specification for an AKS Extension.
+type AKSExtensionSpec struct {
+	Name                    string
+	Namespace               string
+	AKSAssignedIdentityType infrav1.AKSAssignedIdentity
+	ExtensionIdentity       infrav1.ExtensionIdentity
+	AutoUpgradeMinorVersion *bool
+	ConfigurationSettings   map[string]string
+	ExtensionType           *string
+	ReleaseTrain            *string
+	Version                 *string
+	Owner                   string
+	OwnerRef                metav1.OwnerReference
+	Plan                    infrav1.MarketplacePlan
+	Scope                   infrav1.ExtensionScope
+}
+
+// ResourceRef implements azure.ASOResourceSpecGetter.
+func (s *AKSExtensionSpec) ResourceRef() *asokubernetesconfigurationv1.Extension {
+	return &asokubernetesconfigurationv1.Extension{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      s.Name,
+			Namespace: s.Namespace,
+		},
+	}
+}
+
+// Parameters implements azure.ASOResourceSpecGetter.
+func (s *AKSExtensionSpec) Parameters(ctx context.Context, existingAKSExtension *asokubernetesconfigurationv1.Extension) (parameters *asokubernetesconfigurationv1.Extension, err error) {
+	aksExtension := &asokubernetesconfigurationv1.Extension{}
+	if existingAKSExtension != nil {
+		aksExtension = existingAKSExtension
+	}
+
+	aksExtension.Spec = asokubernetesconfigurationv1.Extension_Spec{}
+	aksExtension.Spec.AzureName = s.Name
+	aksExtension.Spec.AutoUpgradeMinorVersion = s.AutoUpgradeMinorVersion
+	aksExtension.Spec.ConfigurationSettings = s.ConfigurationSettings
+	aksExtension.Spec.ExtensionType = s.ExtensionType
+	aksExtension.Spec.ReleaseTrain = s.ReleaseTrain
+	aksExtension.Spec.Version = s.Version
+	aksExtension.Spec.Owner = &genruntime.ArbitraryOwnerReference{
+		ARMID: s.Owner,
+	}
+	aksExtension.Spec.Plan = &asokubernetesconfigurationv1.Plan{
+		Name:      ptr.To(s.Plan.Name),
+		Product:   ptr.To(s.Plan.Product),
+		Publisher: ptr.To(s.Plan.Publisher),
+		Version:   ptr.To(s.Plan.Version),
+	}
+	if s.ExtensionIdentity != "" {
+		aksExtension.Spec.Identity = &asokubernetesconfigurationv1.Identity{
+			Type: (*asokubernetesconfigurationv1.Identity_Type)(ptr.To(s.ExtensionIdentity)),
+		}
+	}
+	if s.AKSAssignedIdentityType != "" {
+		aksExtension.Spec.AksAssignedIdentity = &asokubernetesconfigurationv1.Extension_Properties_AksAssignedIdentity_Spec{
+			Type: (*asokubernetesconfigurationv1.Extension_Properties_AksAssignedIdentity_Type_Spec)(ptr.To(s.AKSAssignedIdentityType)),
```

Comment:
For some reason, the AKSAssignedIdentity type field is not being reflected on Azure. I think it may be overwritten by AKS, but I will need to verify the behavior. In the meantime, I have disabled the check for this field being set in the E2E test: https://github.com/kubernetes-sigs/cluster-api-provider-azure/pull/4360/commits/38eb084ce82692919d0f8249332b70d4e2c5e7f9

Reply:
ping @nojnhuh in case this is an ASO AKS resource edge case

Reply:
I'm not sure what might be going on here besides a general Azure API quirk. @willie-yao  What exactly are you seeing? And how about in the status of the ASO resource?

Reply:
When I describe the extension resource, `AKSAssignedIdentity.type` is null, but the principal ID and tenant ID are populated. I think AKS populates those fields but the user specifies the type as seen [here](https://azure.github.io/azure-service-operator/reference/kubernetesconfiguration/v1api20230501/#kubernetesconfiguration.azure.com/v1api20230501.Extension_Properties_AksAssignedIdentity_Type_Spec)

Reply:
This is what I get with `AKSAssignedIdentity` set to `SystemAssigned` when I run `az k8s-extension show`
```
"aksAssignedIdentity": {
    "principalId": "9899e674-c34c-4bd1-8edf-845e5f5d3f22",
    "tenantId": null,
    "type": null
  },
```